### ISSUE TITLE
[Hard to understand] Form field label is descriptive cc0f0a

### ISSUE THREAD

**Issue by daniel-montalvo** [act-rules/act-rules.github.io#1601]
Rule: https://act-rules.github.io/rules/cc0f0a

What does it mean that "labels are intended for sighted users"? I don't understand this. I think this could be potentially misinterpreted as screen readers necessarily needing some aria-label or aria-labelledby, which I would say it is not the case.

Also I put pr #1600 to fix some typos in the SC names. "Heading and Labels" -> "Headings and Labels".
Please describe the issue with references and relevant examples where necessary.

--- or ---

If authoring a proposal for a new rule, choose from one of the issue templates:

- [Atomic Rule Template](https://raw.githubusercontent.com/act-rules/act-rules.github.io/develop/pages/design/atomic-template-empty.md)
- [Composite Rule Template](https://raw.githubusercontent.com/act-rules/act-rules.github.io/develop/pages/design/composite-template-empty.md)

**Comment by giacomo-petri** [comment]
Considering the context and history of this choice (e.g. https://github.com/act-rules/act-rules.github.io/pull/986), as atomic rule that checks only "visible" labels, we might add the "visible" adjective within the *Description*, making more clear the goal of the rule from the beginning:

From

> Description
> This rule checks that labels describe the purpose of form field elements.

To

> Description
> This rule checks that visible labels describe the purpose of form field elements.

----

On the other hand, applicability section clearly states:

> This rule applies to any [programmatic label](https://act-rules.github.io/rules/cc0f0a#programmatic-label) of an element which has one of the following [semantic roles](https://act-rules.github.io/rules/cc0f0a#semantic-role):
...
and where both the element and the [programmatic label](https://act-rules.github.io/rules/cc0f0a#programmatic-label) are [visible](https://act-rules.github.io/rules/cc0f0a#visible).

so, here we are not talking about the field accessible name, that might differ from the visual presentation of the input label, as per Expectation note:

> Note: Having a [label](https://www.w3.org/TR/WCAG21/#dfn-labels) which is not included in the [accessible name](https://act-rules.github.io/rules/cc0f0a#accessible-name) is a violation of [Success Criterion 2.5.3: Label in Name](https://www.w3.org/WAI/WCAG21/label-in-name) but not of this rule nor of [Success Criterion 2.4.6: Headings and Labels](https://www.w3.org/WAI/WCAG21/#headings-and-labels.html).

----

That said, we might reword

> This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended for sighted users

with something less restrictive, such as:

> This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended to be visually presented to users

**Comment by giacomo-petri** [comment]
Maybe we should also change 

> This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended for sighted users, and that hiding a [visible](https://act-rules.github.io/rules/cc0f0a#visible) [label](https://www.w3.org/TR/WCAG21/#dfn-labels) from assistive technologies, is a failure of [Success Criterion 4.1.2: Name, Role and Value](https://www.w3.org/WAI/WCAG21/#name-role-value), but not of [Success Criterion 2.4.6: Headings and Labels](https://www.w3.org/WAI/WCAG21/#headings-and-labels.html).

in 

> This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended to be visually presented to users, and that hiding a [visible](https://act-rules.github.io/rules/cc0f0a#visible) [label](https://www.w3.org/TR/WCAG21/#dfn-labels) from assistive technologies, is not a failure of [Success Criterion 2.4.6: Headings and Labels](https://www.w3.org/WAI/WCAG21/#headings-and-labels.html), but might be a failure of [Success Criterion 4.1.2: Name, Role and Value](https://www.w3.org/WAI/WCAG21/#name-role-value).

I'm using "might be", because it's not always a 4.1.2 failure, as per following example:
```
<label>
    <span aria-hidden="true">First name</span>
    <input type="text" aria-label="First name" name="fname">
</label>
```

----

Side note:
To clarify that we are referring to the element label and not the accessible name, would be helpful to add the following example to the failures?

```
<label>
    <span>Menu</span>
    <input type="text" aria-label="First name" name="fname">
</label>
```

Per applicability:

> This rule applies to any [programmatic label](https://act-rules.github.io/rules/cc0f0a#programmatic-label) of an element which has one of the following [semantic roles](https://act-rules.github.io/rules/cc0f0a#semantic-role):
> ...
> - textbox
> 
> and where both the element and the [programmatic label](https://act-rules.github.io/rules/cc0f0a#programmatic-label) are [visible](https://act-rules.github.io/rules/cc0f0a#visible).

it's applicable; in fact, per [Examples of Programmatic label](https://act-rules.github.io/pages/examples/programmatic-label/) and [Conflicting label methods](https://act-rules.github.io/pages/examples/programmatic-label/#conflicting-label-methods), both the element and the programmatic label are visible.

Per Expectation:

> Each test target, together with its [visual context](https://act-rules.github.io/rules/cc0f0a#visual-context), describes the purpose of the associated element.
> 
> Note: It is possible for an element to have an [accessible name](https://act-rules.github.io/rules/cc0f0a#accessible-name) but still have a non-descriptive label element (and even a non-descriptive [label](https://www.w3.org/TR/WCAG21/#dfn-labels)). In that case, it would pass [Success Criterion 4.1.2: Name, Role and Value](https://www.w3.org/WAI/WCAG21/#name-role-value) but still fail this rule and [Success Criterion 2.4.6: Headings and Labels](https://www.w3.org/WAI/WCAG21/#headings-and-labels.html).
> 
> Note: Having a [label](https://www.w3.org/TR/WCAG21/#dfn-labels) which is not included in the [accessible name](https://act-rules.github.io/rules/cc0f0a#accessible-name) is a violation of [Success Criterion 2.5.3: Label in Name](https://www.w3.org/WAI/WCAG21/label-in-name) but not of this rule nor of [Success Criterion 2.4.6: Headings and Labels](https://www.w3.org/WAI/WCAG21/#headings-and-labels.html).

As well as failing the 2.4.6 SC, this example is also failing this ACT-rule, because even if the accessible name of the input is meaningful thanks to the aria-label attribute, the visual presentation of the input label is not meaningful (even if someone might say who would do such a thing 🤔😁).

(((Playing a bit devil's advocate, I would argue that this solution is ambiguous to users in general except for users that use assistive technologies; in fact, assistive technologies properly convey the input meaning thanks to the "visually hidden" aria-label property.
Similarly and consistently to 2.4.4: Link Purpose (In Context)), there should be something about "unambiguous in general" also in 2.4.6 SC.)))

----

And allow me one more comment (maybe a bit off topic here, but maybe not): is not so easy to extrapolate information from [2.4.6 SC understanding](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html), which require some effort.
Reading the SC, this is my output:

> This Success Criterion also does not require that content acting as a heading or label be correctly marked up or identified - this aspect is covered separately by [1.3.1: Info and Relationships](https://www.w3.org/WAI/WCAG21/Understanding/info-and-relationships). It is possible for content to pass this Success Criterion (providing descriptive content that acts as headings or labels) while failing Success Criterion 1.3.1 (if the headings or labels aren't correctly marked up/identified). 

is basically saying that 
```
<span>First Name</span>
<input type="text" name="fname">
```
is passing 2.4.6 but not 1.3.1

> Conversely, it is also possible for content to pass Success Criterion 1.3.1 (with headings or labels correctly marked up or identified), while failing this Success Criterion (if those headings or labels are not sufficiently clear or descriptive).
```
<span id="first_name">Menu</span>
<input type="text" name="fname" aria-labelledby="first_name">
```
is passing 1.3.1 but not 2.4.6

> Further, in the case of labels, this Success Criterion does not take into consideration whether or not alternative methods of providing an accessible name for form controls and inputs has been used - this aspect is covered separately by [4.1.2: Name, Role and Value](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value). It is possible for controls and inputs to have an appropriate accessible name (e.g. using aria-label="...") and therefore pass Success Criterion 4.1.2, but to still fail this Success Criterion (if the label is not sufficiently clear or descriptive).

```
<label>
    <span>Address</span>
    <input type="text" aria-label="Apartment" name="apartment_details">
</label>
```
is passing 4.1.2 but not 2.4.6.

So far so clear.

> This success criterion does not require the use of labels; however, it does require that if labels are present, they must be sufficiently clear or descriptive.

On a first reading this might be quite ambiguous. It's clear that labels are not required, but by itself doesn't clearly explain if labels should be visually presented to the user or not, for example if a sr-only label is allowed or not.
```
<label>
    <span class="sr-only">Address</span>
    <input type="text" aria-label="Apartment" name="apartment_details">
</label>
```

By continuing to read and moving to the 2.4.6 success criterion Key Terms, we find the [label](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html#dfn-label) definition:

> **label**
> [text](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html#dfn-text) or other component with a [text alternative](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html#dfn-text-alternative) that is presented to a user to identify a component within Web [content](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html#dfn-content)
> 
> NOTE
> A label is presented to all users whereas the [name](https://www.w3.org/WAI/WCAG21/Understanding/headings-and-labels.html#dfn-name) may be hidden and only exposed by assistive technology. In many (but not all) cases the name and the label are the same.

If my understanding is correct, this note implicitly define the term "label" as "visible label", solving my previous doubt about label visibility.

On the contrary, "A label is presented to all users", means that 
```
<label>
    <span aria-hidden="true">Non meaningful label</span>
    <input type="text" aria-label="Apartment" name="apartment_details">
</label>
```
"Non meaningful label" is not part of the definition, as the aria-hidden="true" prevent AT from informing users about the label, and consequently label is not presented to all users.
Consequently, a "visible label" that has been removed from the accessibility tree seems not fitting into the 2.4.6 criterion applicability/intent/understanding, which is quite inconsistent with my previous "visible label" assumption.
Are my thoughts correct? Is this expected?

Wouldn't be more clear, if my interpretation is correct, to clearly state in 2.4.6 that for this specific criterion, headings and labels are intended to be "visible" and their acc name/order/relationship/etc. are covered by other criteria?

**Comment by dan-tripp-siteimprove** [comment]
If the intended targets of this rule are visual labels only, then shouldn't the applicability be changed away from 'programmatic label' to something else?  I say this because the [ACT examples of 'programmatic label'](https://act-rules.github.io/pages/examples/programmatic-label/) include cases which have an accessible name (via aria-labelledby) and a visual label which is not programmatic.

**Comment by colabottles** [comment]
> That said, we might reword
> 
>     This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended for sighted users
> 
> with something less restrictive, such as:
> 
>     This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended to be visually presented to users

My first inclination from reading through this, is my agreement with @giacomo-petri and "This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended to be visually presented to users." and start from there.

By the wording we have already in the proposed rule, I am understanding it as; 

> This rule checks that labels describe the purpose of form field elements.

To which; 

> This rule applies to any [programmatic label](https://www.w3.org/WAI/standards-guidelines/act/rules/cc0f0a/proposed/#programmatic-label) of an element which has one of the following [semantic roles](https://www.w3.org/WAI/standards-guidelines/act/rules/cc0f0a/proposed/#semantic-role): ...
> 
> and where both the element and the [programmatic label](https://www.w3.org/WAI/standards-guidelines/act/rules/cc0f0a/proposed/#programmatic-label) are [visible](https://www.w3.org/WAI/standards-guidelines/act/rules/cc0f0a/proposed/#visible).

So I'm comprehending this as: "This rule checks that labels describe the the purpose of form field elements. This rule applies to any programmatic label of an element which has one of the following semantic roles" ... "and where both the element and programmatic label are visible and that this rule assumes that labels are intended to be visually presented to users except when not visible and/or not a programmatic label of the element."

I think that if we go too deep this could end up being hard to read for some people who do not understand the guidelines that are looking for explanations as to why this rule exists.

This covers the question from @daniel-montalvo I believe.

> "This rule assumes that [labels](https://www.w3.org/TR/WCAG21/#dfn-labels) are intended to be visually presented to users."

To address @dan-tripp-siteimprove, that also threw me off when originally reading through the proposed rule. I agree that if htis represents "visual labels" then the verbiage should move away from "programmatic label".
